crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,327 @@
+use std::path::PathBuf;
+
+use derive_more::{Deref, DerefMut, Display};
+use flox_types::catalog::cache::{CacheMeta, SubstituterUrl};
+use flox_types::stability::Stability;
+use futures::TryFutureExt;
+use runix::arguments::flake::FlakeArgs;
+use runix::command::Eval;
+use runix::command_line::{NixCommandLine, NixCommandLineRunJsonError};
+use runix::flake_metadata::FlakeMetadata;
+use runix::flake_ref::git::GitRef;
+use runix::flake_ref::indirect::IndirectRef;
+use runix::flake_ref::path::PathRef;
+use runix::flake_ref::{protocol, FlakeRef};
+use runix::installable::{AttrPath, Installable};
+use runix::{RunJson, RunTyped};
+use serde_json::{json, Value};
+use thiserror::Error;
+
+use crate::flox::Flox;
+use crate::providers::git::{GitCommandProvider as Git, GitProvider};
+
+/// Publish state before analyzing
+///
+/// Prevents other actions to commence without analyzing the package first
+pub struct Empty;
+
+/// Publish state after collecting nix metadata
+///
+/// Json value (ideally a [flox_types::catalog::CatalogEntry],
+/// but that's currently broken on account of some flakerefs)
+#[derive(Debug, Deref, DerefMut)]
+pub struct NixAnalysis(Value);
+
+/// State for the publish algorihm
+///
+/// Transitions through typestates to ensure we don't invoke invalid operations
+pub struct Publish<'flox, State> {
+    flox: &'flox Flox,
+    /// The published _upstream_ source
+    publish_ref: PublishRef,
+    /// The published attrpath
+    /// Should be fully resolved to avoid ambiguity
+    attr_path: AttrPath,
+    stability: Stability,
+    analysis: State,
+}
+
+impl<'flox> Publish<'flox, Empty> {
+    pub fn new(
+        flox: &'flox Flox,
+        publish_ref: PublishRef,
+        attr_path: AttrPath,
+        stability: Stability,
+    ) -> Publish<'flox, Empty> {
+        Self {
+            flox,
+            publish_ref,
+            attr_path,
+            stability,
+            analysis: Empty,
+        }
+    }
+
+    /// run analysis on the package and switch to next state
+    ///
+    /// It uses an analyzer flake to extract eval metadata of the derivation.
+    /// The analyzer applies a function to all packages in a `target` flake
+    /// and provides the result under `#analysis.eval.<full attrpath of the package>`.
+    ///
+    /// We evalaute this analysis as json, to which we add
+    /// * source urls for reproducibility
+    /// * the nixpkgs stability being used to create the package
+    pub async fn analyze(self) -> PublishResult<Publish<'flox, NixAnalysis>> {
+        let mut drv_metadata_json = self.get_drv_metadata().await?;
+        let flake_metadata = self.get_flake_metadata().await?;
+
+        // DEVIATION FROM BASH: using `locked` here instead of `resolved`
+        //                      this is used to reproduce the package,
+        //                      but is essentially redundant because of the `source.locked`
+        drv_metadata_json["element"]["url"] = json!(flake_metadata.locked.to_string());
```

Comment:
```suggestion
       // TODO it would be better if we didn't have to do post processing of the analysis for parity with calls to readPackage in https://github.com/flox/floxpkgs/blob/master/modules/common.nix
        drv_metadata_json["element"]["url"] = json!(flake_metadata.locked.to_string());
```

Reply:
It's unfortunate that we need the same logic in floxpkgs in the first place...
If we _do_ need to have this logic within nix (including url substitution etc), 
I'd wish we rather `nix eval -f <the single readPackage> --arg flakeref <flakeref> --arg attrpath '[ "packages" "system" "hello" ]'` or something similar or `nix bundle`.

However, either approach fails to provide the stability override or the original url being used.